kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,234 @@
+///   `&Message<String>` into a `&Message<dyn Display>`. In this case, this would:
+///    - Apply base case to extract the pair `(Message<T>, Message<U>)`.
+///    - Extract the tail element of the struct which are of type `T` and `U`, respectively.
+/// 4. Coercion between smart pointers of wrapper structs.
+///    - Apply the logic from item 2 then item 3.
+pub fn extract_unsize_casting<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    src_ty: Ty<'tcx>,
+    dst_ty: Ty<'tcx>,
+) -> CoercionBase<'tcx> {
+    trace!(?src_ty, ?dst_ty, "extract_unsize_casting");
+    // Iterate over the pointer structure to find the builtin pointer that will store the metadata.
+    let coerce_info = CoerceUnsizedIterator::new(tcx, src_ty, dst_ty).last().unwrap();
+    // Extract the pointee type that is being coerced.
+    let src_pointee_ty = extract_pointee(coerce_info.src_ty).expect(&format!(
+        "Expected source to be a pointer. Found {:?} instead",
+        coerce_info.src_ty
+    ));
+    let dst_pointee_ty = extract_pointee(coerce_info.dst_ty).expect(&format!(
+        "Expected destination to be a pointer. Found {:?} instead",
+        coerce_info.dst_ty
+    ));
+    // Find the tail of the coercion that determines the type of metadata to be stored.
+    let (src_base_ty, dst_base_ty) = tcx.struct_lockstep_tails_erasing_lifetimes(
+        src_pointee_ty,
+        dst_pointee_ty,
+        ParamEnv::reveal_all(),
+    );
+    trace!(?src_base_ty, ?dst_base_ty, "extract_unsize_casting result");
+    assert!(
+        dst_base_ty.is_trait() || dst_base_ty.is_slice(),
+        "Expected trait or slice as destination of unsized cast, but found {dst_base_ty:?}"
+    );
+    CoercionBase { src_ty: src_base_ty, dst_ty: dst_base_ty }
+}
+
+/// This structure represents the base of a coercion.
+///
+/// This base is used to determine the information that will be stored in the metadata.
+/// E.g.: In order to convert an `Rc<String>` into an `Rc<dyn Debug>`, we need to generate a
+/// vtable that represents the `impl Debug for String`. So this type will carry the `String` type
+/// as the `src_ty` and the `dyn Debug` trait as `dst_ty`.
+#[derive(Debug)]
+pub struct CoercionBase<'tcx> {
+    pub src_ty: Ty<'tcx>,
+    pub dst_ty: Ty<'tcx>,
+}
+
+/// Iterates over the coercion path of a structure that implements `CoerceUnsized<T>` trait.
+/// The `CoerceUnsized<T>` trait indicates that this is a pointer or a wrapper for one, where
+/// unsizing can be performed on the pointee. More details:
+/// <https://doc.rust-lang.org/std/ops/trait.CoerceUnsized.html>
+///
+/// Given an unsized coercion between `impl CoerceUnsized<T>` to `impl CoerceUnsized<U>` where
+/// `T` is sized and `U` is unsized, this iterator will walk over the fields that lead to a
+/// pointer to `T`, which shall be converted from a thin pointer to a fat pointer.
+///
+/// This iterator will also include the name of the field that differs in the current iteration
```

Comment:
Isn't it the `Info` not the `iterator` that has the field name? So we should really say "the result of iterating with this iterator will ..."

Reply:
Indeed. Let me improve that.